File: 算法/代码随想录算法PDF/00调试源代码/257.二叉树的所有路径.py
#
# @lc app=leetcode.cn id=257 lang=python3
#
# [257] 二叉树的所有路径
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        res = []
        if not root: return res
        treeSt, pathSt = [root], [str(root.val)]
        while treeSt:
            node = treeSt.pop()
            path = pathSt.pop()
            if not node.left and not node.right:
                res.append(path)
            if node.right:
                treeSt.append(node.right)
                pathSt.append(path + "->" + str(node.right.val))
            if node.left:
                treeSt.append(node.left)
                pathSt.append(path + "->" + str(node.left.val))
        return res
# @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/257.二叉树的所有路径.cpp
/*
 * @lc app=leetcode.cn id=257 lang=cpp
 *
 * [257] 二叉树的所有路径
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
     vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> res;
        if (!root) return res;
        stack<TreeNode*> treeSt;
        stack<string> pathSt;
        treeSt.push(root);
        pathSt.push(to_string(root->val));
        while (!treeSt.empty()) {
            TreeNode* node = treeSt.top();
            treeSt.pop();
            string path = pathSt.top();
            pathSt.pop();
            if (!node->left && !node->right) {
                res.push_back(path);
            }
            if (node->left) {
                treeSt.push(node->left);
                pathSt.push(path + "->" + to_string(node->left->val));
            }
            if (node->right) {
                treeSt.push(node->right);
                pathSt.push(path + "->" + to_string(node->right->val));
            }
        }
        return res;
    }
};
// @lc code=end

